pyanaconda/modules/users/users.py

```diff
@@ -84,58 +181,91 @@ def generate_kickstart(self, skip_unsupported=False):
         data.rootpw.lock = self.root_account_locked
         data.rootpw.seen = self.rootpw_seen
 
-        if skip_unsupported:
-            return str(data)
+        for user_data in self.users:
+            data.user.userList.append(user_data_to_ksdata(user_data))
+
+        for group_data in self.groups:
+            group_ksdata = GroupKickstartData()
+            group_ksdata.name = group_data.name
+            if group_data.gid == -1:
+                group_data.gid = None
+            else:
+                group_ksdata.gid = group_data.gid
+            data.group.groupList.append(group_ksdata)
 
-        for user in self.users.values():
-            user.setup_kickstart(data)
+        for ssh_key_data in self.ssh_keys:
+            ssh_key_ksdata = SshKeyKickstartData()
+            ssh_key_ksdata.key = ssh_key_data.key
+            ssh_key_ksdata.username = ssh_key_data.username
+            data.sshkey.sshUserList.append(ssh_key_ksdata)
 
         return str(data)
 
+    def install_with_tasks(self, sysroot):
+        """Return the installation tasks of this module.
+
+        :param str sysroot: a path to the root of the installed system
+        :returns: list of object paths of installation tasks
+        """
+        tasks = [
+            SetRootPasswordTask(sysroot=sysroot, password=self.root_password,
+                                crypted=self.root_password_is_crypted,
+                                locked=self.root_account_locked),
+            CreateGroupsTask(sysroot=sysroot, group_data_list=self.groups),
+            CreateUsersTask(sysroot=sysroot, user_data_list=self.users),
+            SetSshKeysTask(sysroot=sysroot, ssh_key_data_list=self.ssh_keys)
+        ]
+
+        paths = [
+            self.publish_task(USERS.namespace, task) for task in tasks
+        ]
+
+        return paths
+
     @property
     def users(self):
-        """Dictionary of users and their object paths."""
+        """List of DBUS structures, one per user."""
         return self._users
 
-    @property
-    def object_paths_of_users(self):
-        """List of users object paths."""
-        return list(self._users.keys())
-
-    def create_user(self):
-        """Create and publish a new UserModule.
+    def set_users(self, users):
+        """Set the list of DBUS structures, one per user."""
+        self._users = users
+        self.users_changed.emit()
+        log.debug("A new user list has been set (%d users).", len(self._users))
```

Comment:
We are trying to always log the new values to make the debugging simpler in the future. Please, define the `__str__` methods of your data classes and log the data objects instead.

Reply:
Yeah, that makes sense. Initially I simply wanted to avoid the password ending up in logs, but that should, hopefully be the only thing. Theoretically people could have something sensitive in other fields (GECOS ?), but hardly as sensitive as passwords.

Reply:
Implemented.